### PR TITLE
KAFKA-6161 Add default implementation to close() and configure() for …

### PR DESCRIPTION
**Pull request by chia7712**
### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by saidbouras** [comment]
I am looking into that PR, and I have a question. 
Now that `Serializer` and `Deserializer` are functional interfaces, we would implement only methods : `serialize` and `deserialize` and if we want to configure we have to override the parent method. I think it's a bit odd, i'd rather add and interfaces that extends thoses (we can call them `UnConfigurableSerializer`, `UnConfigurableDeserializer`), it would be more natural and we can directly know what we are doing by implementing them.

**Comment by chia7712** [comment]
@saidbouras Could you add the idea to thread in order to discuss with whole community?
https://lists.apache.org/thread.html/b7e7387fc5523a31150aaffc221d983cfd7958a04540c2b1715d7e7a@%3Cdev.kafka.apache.org%3E

**Comment by guozhangwang** [comment]
cc @cmccabe @ijuma @hachikuji

**Comment by mjsax** [comment]
@chia7712 Can you also update the docs?

**Comment by chia7712** [comment]
> Can you also update the docs?

Pardon me. Do you mean the docs of close() and configurate()?

**Comment by mjsax** [comment]
Sorry for being unclear. I mean the "upgrade docs", in particular `docs/streams/upgrade-guide.html` -- for each release, we have a section for public API changes.

Compare #6108 that is just adding this section (maybe it's best to wait until the PR is merged and rebase your PR afterwards).

**Comment by chia7712** [comment]
>  I mean the "upgrade docs", in particular docs/streams/upgrade-guide.html -- for each release, we have a section for public API changes.

Copy that.

**Comment by chia7712** [comment]
> Compare #6108 that is just adding this section (maybe it's best to wait until the PR is merged and rebase your PR afterwards).

Got it

**Comment by chia7712** [comment]
I overlooked the build for scala 2.11...Will fix the compilation error

File: clients/src/main/java/org/apache/kafka/common/serialization/Deserializer.java
Comment (mjsax) [inline clients/src/main/java/org/apache/kafka/common/serialization/Deserializer.java]: If you want a new paragraph, you need to insert a `<p>` instead of an empty line.

File: clients/src/main/java/org/apache/kafka/common/serialization/Serde.java
Comment (mjsax) [inline clients/src/main/java/org/apache/kafka/common/serialization/Serde.java:44]: Maybe insert `<p>` to split into two paragraphs as above?

File: clients/src/main/java/org/apache/kafka/common/serialization/Serializer.java
Comment (mjsax) [inline clients/src/main/java/org/apache/kafka/common/serialization/Serializer.java:68]: As above.

File: docs/streams/upgrade-guide.html
Comment (mjsax) [inline docs/streams/upgrade-guide.html]: This goes into `2.3.0` now.

File: streams/src/main/java/org/apache/kafka/streams/Topology.java
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/Topology.java:307]: Why do we need this? Does the build report a WARNING and fails without it? (same below)

File: streams/test-utils/src/test/java/org/apache/kafka/streams/TopologyTestDriverTest.java
Comment (mjsax) [inline streams/test-utils/src/test/java/org/apache/kafka/streams/TopologyTestDriverTest.java]: Nit: keep first two parameters in their own lines (we either put all parameters in one line, or use one parameter per line -- hybrid formatting makes it harder to read the code).

**Review by mjsax — COMMENTED** [review]
Thanks for updating the PR. Overall LGTM. Just a few minor comments.

File: streams/src/main/java/org/apache/kafka/streams/Topology.java
Comment (chia7712) [reply streams/src/main/java/org/apache/kafka/streams/Topology.java:307]: Adding default implementation to close() and configure() make Deserializer be a functional interface. Hence, it is feasible to pass lambdas without explicit type (Deserializer and TimestampExtractor can be the identical lambda expression). javac doesn't like the potential ambiguity. We can reorder the arguments in order to "avoid" the complaint. However, Topology is a public class so changing the method signature will break the compatibility. As a result, I added SuppressWarnings.

Comment (mjsax) [reply streams/src/main/java/org/apache/kafka/streams/Topology.java:307]: Thanks for the details -- however, you did not answer my question: Does the build fail without the annotation?

Comment (chia7712) [reply streams/src/main/java/org/apache/kafka/streams/Topology.java:307]: > Does the build fail without the annotation?

absolutely :)

**Review by mjsax — APPROVED** [review]
LGTM.

Call for second review @guozhangwang @hachikuji

File: docs/streams/upgrade-guide.html
Comment (guozhangwang) [inline docs/streams/upgrade-guide.html]: This is under the

```
<h3><a id="streams_api_changes_220" href="#streams_api_changes_220">Streams API changes in 2.2.0</a></h3>
```

Is that intentional?

I think we should create new `Streams API changes in 2.3.0` section on top and add this paragraph there.

**Review by guozhangwang — APPROVED** [review]
Left a comment about docs change, otherwise LGTM.

@mjsax please feel free to merge after it is addressed.

File: docs/streams/upgrade-guide.html
Comment (chia7712) [reply docs/streams/upgrade-guide.html]: > I think we should create new `Streams API changes in 2.3.0` section on top and add this paragraph there.

Thanks for reviews. I will address this comment by next commit.

**Comment by chia7712** [comment]
failed tests - AdminClientIntegrationTest and DynamicBrokerReconfigurationTest - pass locally.

**Comment by chia7712** [comment]
retest this please

**Comment by chia7712** [comment]
> 07:41:07 org.apache.kafka.streams.kstream.internals.KTableKTableLeftJoinTest > shouldNotThrowIllegalStateExceptionWhenMultiCacheEvictions STARTED
08:31:52 
08:31:52 org.apache.kafka.streams.kstream.internals.KTableKTableLeftJoinTest > shouldNotThrowIllegalStateExceptionWhenMultiCacheEvictions PASSED

not sure why QA is frozen...will dig in it.

**Comment by chia7712** [comment]
The other PRs have similar issue. For example, https://builds.apache.org/job/kafka-pr-jdk11-scala2.12/2408/consoleFull, https://builds.apache.org/job/kafka-pr-jdk11-scala2.12/2413/consoleFull and https://builds.apache.org/job/kafka-pr-jdk11-scala2.12/2412/consoleFull.

@mjsax I believe the frozen QA is unrelated to this PR. Should I retry QA again?

**Comment by mjsax** [comment]
It's a known issue: https://issues.apache.org/jira/browse/KAFKA-7933

Retest this please.

**Comment by chia7712** [comment]
Retest this please.

**Comment by chia7712** [comment]
shouldNotThrowIllegalStateExceptionWhenMultiCacheEvictions is still slow.

**Comment by mjsax** [comment]
Retest this please.

**Comment by chia7712** [comment]
shouldNotThrowIllegalStateExceptionWhenMultiCacheEvictions is fixing by #6292

Will retry the QA if the PR is merged...

**Comment by chia7712** [comment]
DeleteConsumerGroupsTest.testDeleteCmdWithShortInitialization pass on local

**Comment by guozhangwang** [comment]
Thank you @chia7712 !

**Pull request merged**
